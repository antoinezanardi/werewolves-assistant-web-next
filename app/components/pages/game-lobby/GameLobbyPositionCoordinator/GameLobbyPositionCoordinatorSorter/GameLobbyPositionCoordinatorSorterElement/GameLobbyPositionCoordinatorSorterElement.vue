<template>
  <li
    id="game-lobby-position-coordinator-sorter-element"
    :key="player.name"
    class="border-2 border-gray-700 cursor-move flex gap-2 items-center rounded-md shadow-black shadow-md w-full"
  >
    <div
      id="position-coordinator-sorter-element-square"
      class="h-8 w-8"
      :style="{ 'background-color': playerColor }"
    />

    <RoleImage
      id="position-coordinator-sorter-element-role-image"
      class="h-8 w-8"
      definition="small"
      :role-name="player.role.name"
    />

    <div
      id="position-coordinator-sorter-element-player-name"
      class="grow truncate"
    >
      {{ player.name }}
    </div>

    <i class="fa fa-bars me-2"/>
  </li>
</template>

<script setup lang="ts">
import stringToColor from "string-to-color";
import type { GameLobbyPositionCoordinatorSorterElementProps } from "~/components/pages/game-lobby/GameLobbyPositionCoordinator/GameLobbyPositionCoordinatorSorter/GameLobbyPositionCoordinatorSorterElement/game-lobby-position-coordinator-sorter-element.types";
import RoleImage from "~/components/shared/role/RoleImage/RoleImage.vue";

const props = defineProps<GameLobbyPositionCoordinatorSorterElementProps>();

const playerColor = computed<string>(() => stringToColor(props.player.name));
</script>