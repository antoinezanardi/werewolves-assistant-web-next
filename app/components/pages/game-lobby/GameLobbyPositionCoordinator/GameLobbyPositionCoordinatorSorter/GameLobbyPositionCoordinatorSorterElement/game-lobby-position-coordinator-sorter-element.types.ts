import type { CreateGamePlayerDto } from "~/composables/api/game/dto/create-game/create-game-player/create-game-player.dto";

type GameLobbyPositionCoordinatorSorterElementProps = {
  player: CreateGamePlayerDto;
};

export type { GameLobbyPositionCoordinatorSorterElementProps };