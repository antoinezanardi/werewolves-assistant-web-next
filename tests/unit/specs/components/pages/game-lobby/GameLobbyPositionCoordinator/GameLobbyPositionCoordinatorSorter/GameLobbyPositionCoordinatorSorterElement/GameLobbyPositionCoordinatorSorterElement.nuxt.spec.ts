import type { mount } from "@vue/test-utils";
import type { ComponentMountingOptions } from "@vue/test-utils/dist/mount";

import { createFakeCreateGamePlayerDto } from "@tests/unit/utils/factories/composables/api/game/dto/create-game/create-game-player/create-game-player.dto.factory";
import { mountSuspendedComponent } from "@tests/unit/utils/helpers/mount.helpers";
import type { GameLobbyPositionCoordinatorSorterElementProps } from "~/components/pages/game-lobby/GameLobbyPositionCoordinator/GameLobbyPositionCoordinatorSorter/GameLobbyPositionCoordinatorSorterElement/game-lobby-position-coordinator-sorter-element.types";
import GameLobbyPositionCoordinatorSorterElement from "~/components/pages/game-lobby/GameLobbyPositionCoordinator/GameLobbyPositionCoordinatorSorter/GameLobbyPositionCoordinatorSorterElement/GameLobbyPositionCoordinatorSorterElement.vue";
import type RoleImage from "~/components/shared/role/RoleImage/RoleImage.vue";

describe("Game Lobby Position Coordinator Sorter Element Component", () => {
  const defaultProps: GameLobbyPositionCoordinatorSorterElementProps = {
    player: createFakeCreateGamePlayerDto({
      name: "Player 1",
      role: { name: "seer" },
    }),
  };
  let wrapper: ReturnType<typeof mount<typeof GameLobbyPositionCoordinatorSorterElement>>;

  async function mountGameLobbyPositionCoordinatorSorterElementComponent(options: ComponentMountingOptions<typeof GameLobbyPositionCoordinatorSorterElement> = {}):
  Promise<ReturnType<typeof mount<typeof GameLobbyPositionCoordinatorSorterElement>>> {
    return mountSuspendedComponent(GameLobbyPositionCoordinatorSorterElement, {
      props: defaultProps,
      ...options,
    });
  }

  beforeEach(async() => {
    wrapper = await mountGameLobbyPositionCoordinatorSorterElementComponent();
  });

  it("should match snapshot when rendered.", () => {
    expect(wrapper).toBeTruthy();
    expect(wrapper.html()).toMatchSnapshot();
  });

  describe("Player Square", () => {
    it("should set background color to square depending on player name when rendered.", () => {
      const playerSquare = wrapper.find<HTMLDivElement>("#position-coordinator-sorter-element-square");

      expect(playerSquare.attributes("style")).toBe("background-color: #daa42b;");
    });
  });

  describe("Role Image", () => {
    it("should set player role name to role image when rendered.", () => {
      const roleImage = wrapper.findComponent<typeof RoleImage>("#position-coordinator-sorter-element-role-image");

      expect(roleImage.props("roleName")).toBe("seer");
    });
  });

  describe("Player Name", () => {
    it("should set player name to player name text when rendered.", () => {
      const playerName = wrapper.find<HTMLSpanElement>("#position-coordinator-sorter-element-player-name");

      expect(playerName.text()).toBe("Player 1");
    });
  });
});