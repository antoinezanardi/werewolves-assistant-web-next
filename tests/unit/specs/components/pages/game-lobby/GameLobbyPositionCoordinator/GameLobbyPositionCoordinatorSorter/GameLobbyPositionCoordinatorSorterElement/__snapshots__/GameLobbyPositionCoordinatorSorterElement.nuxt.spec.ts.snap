// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Game Lobby Position Coordinator Sorter Element Component > should match snapshot when rendered. 1`] = `
"<li id="game-lobby-position-coordinator-sorter-element" class="border-2 border-gray-700 cursor-move flex gap-2 items-center rounded-md shadow-black shadow-md w-full">
  <div id="position-coordinator-sorter-element-square" class="h-8 w-8" style="background-color: #daa42b;"></div>
  <role-image-stub rolename="seer" sizes="50px" definition="small" id="position-coordinator-sorter-element-role-image" class="h-8 w-8"></role-image-stub>
  <div id="position-coordinator-sorter-element-player-name" class="grow truncate">Player 1</div><i class="fa fa-bars me-2"></i>
</li>"
`;
